### PR TITLE
reorgs implementation

### LINKED ISSUE #51: unittest reorg function
The class of reorg detection need to migrate the test from node and add new one for the extra parameter

### PR DESCRIPTION
**Pull request by joanestebanr**
Closes #51, #49, #29 , #28, #27, #26, #21

## Interface changes
Add new function to get L1Blocks:
```
GetLastL1Block(ctx context.Context) (*L1Block, error)
GetL1BlockByNumber(ctx context.Context, blockNumber uint64) (*L1Block, error)
```

Config changes: 
```
[Synchronizer]
	(...)
	SyncBlockProtection = "latest"
	SyncBlockProtectionOffset = -5
```
**SyncBlockProtection** and **SyncBlockProtectionOffset** define the synchronization point from L1. By default is 5 block before the `latest`. 
Before this change was `latest` always. 

### Reviewers

Main reviewers:

- @John
- @Doe

Codeowner reviewers:

- @-Alice
- @-Bob

### CODE REVIEW

File: state/storage/interface.go [hunk 1]
```diff
@@ -16,9 +16,10 @@ type storageTxType = entities.Tx
 
 type BlockStorer interface {
 	AddBlock(ctx context.Context, block *L1Block, dbTx storageTxType) error
+	UpdateCheckedBlockByNumber(ctx context.Context, blockNumber uint64, newCheckedStatus bool, dbTx storageTxType) error
 	GetLastBlock(ctx context.Context, dbTx storageTxType) (*L1Block, error)
 	GetBlockByNumber(ctx context.Context, blockNumber uint64, dbTx storageTxType) (*L1Block, error)
-	GetPreviousBlock(ctx context.Context, offset uint64, dbTx storageTxType) (*L1Block, error)
+	GetPreviousBlock(ctx context.Context, offset uint64, fromBlockNumber *uint64, dbTx storageTxType) (*L1Block, error)
```
Comment (ARR552) [inline state/storage/interface.go]: I don't understand the need of this change

File: state/storage/pgstorage/block.go [hunk 2]
```diff
@@ -30,14 +43,21 @@ func (p *PostgresStorage) GetBlockByNumber(ctx context.Context, blockNumber uint
...
+		whereClause = "WHERE block_num <= " + fmt.Sprintf("%d", *fromBlockNumber)
+	}
+	getPreviousBlockSQL := "SELECT block_num, block_hash, parent_hash, received_at,checked,has_events,sync_version FROM sync.block " + whereClause + " ORDER BY block_num DESC LIMIT 1 OFFSET $1"
 	return p.queryBlock(ctx, fmt.Sprintf("GetPreviousBlock %d", offset), getPreviousBlockSQL, dbTx, offset)
 }
 
 // GetFirstUncheckedBlock returns the first L1 block that has not been checked from a given block number.
 func (p *PostgresStorage) GetFirstUncheckedBlock(ctx context.Context, fromBlockNumber uint64, dbTx dbTxType) (*L1Block, error) {
-	const getLastBlockSQL = "SELECT block_num, block_hash, parent_hash, received_at, has_events,checked FROM sync.block  WHERE block_num>=$1 AND  checked=false ORDER BY block_num LIMIT 1"
+	const getLastBlockSQL = "SELECT block_num, block_hash, parent_hash, received_at,checked,has_events,sync_version FROM sync.block  WHERE block_num>=$1 AND  checked=false ORDER BY block_num LIMIT 1"
```
Comment (ARR552) [inline state/storage/pgstorage/block.go:56]: I don't understand neither has_event param nor sync_version. Could you explain them?

File: state/storage/pgstorage/kv.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package pgstorage
```
Comment (ARR552) [inline state/storage/pgstorage/kv.go:1]: What is the purpose of this? I think this is not related to this PR. If so, could you rip if off?

File: state/storage/pgstorage/migrations/0004.sql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+-- +migrate Up
+CREATE TABLE IF NOT EXISTS sync.kv (
+    key VARCHAR(256) PRIMARY KEY,
+    value VARCHAR
+);
```
Comment (ARR552) [inline state/storage/pgstorage/migrations/0004.sql]: What is the relationship between getBlock and this??
Why kv is needed?

File: state/storage/pgstorage/pgstorage_test.go [hunk 5]
```diff
@@ -46,63 +45,6 @@ func TestQueries(t *testing.T) {
 	require.Equal(t, sb.FromBatchNumber, data.FromBatchNumber)
 }
 
-func TestBlockAddAndGets(t *testing.T) {
```
Comment (ARR552) [inline state/storage/pgstorage/pgstorage_test.go:49]: Can you update the test instead just deleting it?

File: state/storage/pgstorage/tx.go [hunk 6]
```diff
@@ -3,25 +3,42 @@ package pgstorage
...
+	entities.Tx
 }
 
+type TxCallbackType = entities.TxCallbackType
+
 type stateImplementationTx struct {
 	pgx.Tx
-	rollbackCallbacks []func()
+	rollbackCallbacks []TxCallbackType
+	commitCallbacks   []TxCallbackType
```
Comment (ARR552) [inline state/storage/pgstorage/tx.go:20]: I don't understand the purpose of this

File: synchronizer/common/syncinterfaces/state.go [hunk 7]
```diff
@@ -21,6 +21,11 @@ type stateOnSequencedBatchesManager interface {
 	OnSequencedBatchesOnL1(ctx context.Context, seq model.SequenceOfBatches, dbTx stateTxType) error
 }
 
+type stateReorgManager interface {
+	AddOnReorgCallback(f model.ReorgCallbackType)
+	ExecuteReorg(ctx context.Context, reorgRequest model.ReorgRequest, dbTx stateTxType) model.ReorgExecutionResult
+}
```
Comment (ARR552) [inline synchronizer/common/syncinterfaces/state.go:27]: Finally you split the reorg logic to another class...

File: synchronizer/l1_sync/block_range_iterator.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
+package l1sync
```
Comment (ARR552) [inline synchronizer/l1_sync/block_range_iterator.go:1]: Do you think it is worth it to do it in this way?? 5 line of code to maintain vs 80 lines

File: synchronizer/l1_sync/check_reorgs.go [hunk 9]
```diff
@@ -135,8 +148,8 @@ func (s *CheckReorgManager) NewCheckReorg(latestStoredBlock *stateBlockType, syn
 	if latestStoredEthBlock.BlockHash != reorgedBlock.BlockHash {
 		latestStoredBlock = &reorgedBlock
 		log.Info("Reorg detected in block: ", latestStoredEthBlock.BlockNumber, " last block OK: ", latestStoredBlock.BlockNumber)
-		return latestStoredBlock, nil
+		return latestStoredBlock, latestStoredEthBlock.BlockNumber, nil
 	}
 	log.Debugf("No reorg detected in block: %d. BlockHash: %s", latestStoredEthBlock.BlockNumber, latestStoredEthBlock.BlockHash.String())
-	return nil, nil
+	return nil, 0, nil
```
Comment (ARR552) [inline synchronizer/l1_sync/check_reorgs.go:143]: Doesn't make any sense to return a value that it is included as input param (`latestStoredEthBlock`)

File: synchronizer/l1_sync/l1_sequential_sync.go [hunk 10]
```diff
@@ -2,158 +2,318 @@ package l1sync
...
+		if err != nil {
+			log.Warnf("error processing the block range. Err: %v", err)
+			return nil, err
+		}
+		return entities.NewL1BlockFromEthermanBlock(emptyBlock, true), nil
+	}
+	return nil, nil
 }
 
+/*
```
Comment (ARR552) [inline synchronizer/l1_sync/l1_sequential_sync.go]: remove this code

File: synchronizer/state_tx_reuse.go [hunk 11]
```diff
@@ -2,42 +2,55 @@ package synchronizer
...
+// object: ReuseStateTxProvider
+// intent:
+//   for a process that create a new tx each time
+//   you can bypass this behavious returning a 'fake tx' ReuseStateTx
+//   that ignore commit. The rollback is real
+//   The usage implies that the level that use this
+//   will commit / rollback the real Tx
+// It substitute the state class that implements
+//    BeginTransaction
+
```
Comment (ARR552) [inline synchronizer/state_tx_reuse.go]: can you explain i which cases this is needed? Seems a bit dangerous

**Review by ARR552 — CHANGES_REQUESTED** [review]
I don't agree with many changes.These changes will over complicate the code, for instance creating unneeded classes everywhere. @arnaubennassar faced these problems some weeks ago.
Over complicate the code it is not a good way to go. Keeping the code as much simple us possible reduces bugs and allows others to contribute easily

File: synchronizer/state_tx_reuse.go (hunk 11 above)
Comment (joanestebanr) [reply synchronizer/state_tx_reuse.go]: This is used to sync  events previous to genesis (needs for other rollups that starts after LxLy deployment). It doesn't store l1block so it need to be atomic to be sure that have been executed totally

File: state/storage/pgstorage/block.go (hunk 2 above)
Comment (joanestebanr) [reply state/storage/pgstorage/block.go:56]: Yes for sure: 
This implementation maintain the 'empty blocks' as bookmarks of synchronization point
-  **has_event**: means that this block contains rollup events, if it's `false` means that is just a check point
- **sync_version**: is to keep a track of each block which version create it, so if there is a bug we can decide easly which blocks are affected

File: state/storage/pgstorage/kv.go (hunk 3 above)
Comment (joanestebanr) [reply state/storage/pgstorage/kv.go:1]: This is part of the WIP. The idea is create a Key/Value store to some values as:
- The chainID  that was synced in this DB
- The RollupID that was synced in this DB 
and so on...

File: state/storage/pgstorage/migrations/0004.sql (hunk 4 above)
Comment (joanestebanr) [reply state/storage/pgstorage/migrations/0004.sql]: - No relation between getBlock and this: this PR is a partial implementation of reorg feature
- I explain the use of *kv* in the previous comment

File: state/storage/pgstorage/pgstorage_test.go (hunk 5 above)
Comment (joanestebanr) [reply state/storage/pgstorage/pgstorage_test.go:49]: It's not delete, just moved to `block_test.go`

File: synchronizer/common/syncinterfaces/state.go (hunk 7 above)
Comment (joanestebanr) [reply synchronizer/common/syncinterfaces/state.go:27]: Yes: 
- It makes so much easy unittest
- It makes more easy to understand (even keeping the same long function)
- It respects the SRP principle

File: synchronizer/l1_sync/block_range_iterator.go (hunk 8 above)
Comment (joanestebanr) [reply synchronizer/l1_sync/block_range_iterator.go:1]: This class have the logic of iterating over the blocks from the legacy code. The legacy code was coupled to the main loop and makes a bit difficult to follow how it's work.  
Also have included a modification that is allow to GetRange overlapped (for reorg check) or not overlapped if it doesn't need that 
- The code is easy to understand
- The code is easy to check

File: synchronizer/l1_sync/check_reorgs.go (hunk 9 above)
Comment (joanestebanr) [reply synchronizer/l1_sync/check_reorgs.go:143]: Sorry, I don't understand that.. do you mean that I received as param `latestStoredBlock` and returns the same concept?

**Comment by joanestebanr** [comment]
> I don't agree with many changes.These changes will over complicate the code, for instance creating unneeded classes everywhere. @arnaubennassar faced these problems some weeks ago. Over complicate the code it is not a good way to go. Keeping the code as much simple us possible reduces bugs and allows others to contribute easily

First of all, I want to thank @ARR552  for the time invested, since it is pretty big PR. It has the reorg implementation but also some internal refactors. As you know, the base code used is a mix between node and bridge, so it has significant technical debt. Also take in account that is partial PR so, the clean stage is going to be done at the end of the work

I totally understand your concerns. For sure, the goal is to create the best code possible, and one of the qualities is to be simple. But creating interfaces and dividing the code into small pieces is a good way to achieve a good code. It allows you to test it more easily, but on the other hand, it's true that it complicates the structure. Creating interfaces and objects, we get a better isolation of code, and it allows to unitest   

This code follows clean code and  SOLID principles, with more or less fortune, increasing the cohesion of classes, reducing coupling, reducing cyclomatic complexity  and avoiding some antipatterns like spaghetti code. 

I agree that we can improve it,  reducing some parts or re-thinking about new solutions that fit better. So, I propose meeting to discuss each case and find the best trade-off.

File: state/storage/interface.go (hunk 1 above)
Comment (joanestebanr) [reply state/storage/interface.go]: You are right, I use it in some old code but right now is no longer needed! Good catch!

File: state/storage/pgstorage/block.go (hunk 2 above)
Comment (ARR552) [reply state/storage/pgstorage/block.go:56]: But if you keep the empty block is the same behavior that lead to the problems we had and will fail again

File: synchronizer/l1_sync/check_reorgs.go (hunk 9 above)
Comment (joanestebanr) [reply synchronizer/l1_sync/check_reorgs.go:143]: After the meeting I got the idea, I will migrate the unittest and add news to check the second returned value #51

File: state/storage/pgstorage/tx.go (hunk 6 above)
Comment (joanestebanr) [reply state/storage/pgstorage/tx.go:20]: This is the generalization of the fix for the issue of L1 infotree leaf mismatch in the node (https://github.com/0xPolygonHermez/zkevm-node/issues/3200)

**Review by ARR552 — APPROVED** [review]
